### ISSUE TITLE
Add config option for the inferno "base" URL

### ISSUE THREAD

**Issue by cooperthompson** [onc-healthit/inferno-program#317]
**Is your feature request related to a problem? Please describe it.**
Our system requires that redirect URLs be "https".  We also strongly encourage https for EHR launch URLs (and behave differently for https vs. http launch URLs).

Because the Inferno nginx server runs http, I have set up an Apache web server to proxy traffic:

<VirtualHost *:443>
        ServerAdmin name@example.com
        ServerName inferno.epic.com
        ErrorLog ${APACHE_LOG_DIR}/inferno-error.log
        CustomLog ${APACHE_LOG_DIR}/inferno-access.log combined

        SSLEngine on
        SSLCertificateFile /path/to/ssl/inferno.cer
        SSLCertificateKeyFile /path/to/ssl/private/inferno.key
        SSLCertificateChainFile /path/to/ssl/certs/InternalCAFile.pem

        LogLevel warn
        ProxyPass / http://inferno.epic.com:4567/
        ProxyPassReverse / https://inferno.epic.com:4567/

</VirtualHost>

This mostly works great.  However, when Inferno runs several of the tests, it pre-populates the redirect and launch URLs based on (I think) the request.base_url.  Because of the proxy, it picks up http instead of https, and uses that during the OAuth2 flows.

Note:  I'm a little confused on why it doesn't pick up the port 4567, but I haven't looked in to how Ruby deals with request.base_url.

**Describe the solution you'd like to see implemented**
If we had a config.yml option to specify the external base URL of the inferno server, then I could specify the Apache-fronted https base URL there, and ideally Inferno would use that to generate the URLs it uses launch, redirect, and JWK URLs it uses.

Alternatively, if there is something I'm missing, and my https fronting proxy should be doing something different to get Inferno to pick up the https, let me know.

**Describe alternatives you've considered**
Describe any alternative solutions or features you have considered.
I think the main alternative is to update nginx to host on https, but that involves setting up a cert and I don't know nginx or containers all that well.

**Comment by Jammjammjamm** [comment]
As a quick workaround, the configuration for our deployed instances (which use https) lives here: https://github.com/onc-healthit/inferno.healthit.gov

That repo deploys more than just program, so you could either comment out the services you don't need from that docker-compose and have it load your certs, or use its nginx configuration as a model for yours.

**Comment by arscan** [comment]
You can see how we configured the (docker-compose internal) nginx server to communicate the fact that the Ruby server is running behind SSL in our production environment (inferno.healthit.gov, which has more than just "inferno-program" running) here: https://github.com/onc-healthit/inferno.healthit.gov/blob/2a3a4290f16c801bc40347b1858fb0cad537cad1/nginx/nginx.conf#L69-L72

Looks like we set some (defacto) standard headers that Ruby/Rack/Sinatra picks up on: `X-Forwarded-Proto` and `X-Forwarded-Port`.  We also re-write the `Host` header it seems.

So we just need to make sure those headers make it to Ruby (either provided by Apache and passed through nginx, or just hardcoded in nginx).  If Apache is generating them already, it may not work because it doesn't look like we have added `proxy_pass_request_headers      on;` to this section: https://github.com/onc-healthit/inferno-program/blob/c66e424c41b4dcc04bcfffe9f9621e534db6f6f6/docker/nginx.conf#L62-L70

So you could add that line to your copy and restart the nginx docker instance you are running (if Apache is providing the forwarded headers), I think.

Or, if you are unsure whether or not Apache is doing that at all, and know that you will only run this in SSL, then I think you could just add `proxy_set_header X-Forwarded-Proto https;` to that section instead https://github.com/onc-healthit/inferno-program/blob/c66e424c41b4dcc04bcfffe9f9621e534db6f6f6/docker/nginx.conf#L62-L70

Does that help?  I think we'll probably want to update this file so it better handles being put in front of a reverse proxy (at least by passing through the forwarded proto/port headers).

**Comment by cooperthompson** [comment]
I'd actually figured out most of this in parallel.  I posted on the [chat.fhir thread](https://chat.fhir.org/#narrow/stream/179309-inferno/topic/Run.20inferno.20Docker.20image.20with.20https.20redirect.20URI), but the key setting for me was:

proxy_set_header X-Forwarded-Proto https;

**Comment by yunwwang** [comment]
@cooperthompson I will close this ticket as it has been address. If the problem is still there, please either reopen this ticket or create a new issue.
